mirascope/core/base/prompt.py

```diff
@@ -298,19 +305,67 @@ async def run():
 
 
 _BasePromptT = TypeVar("_BasePromptT", bound=BasePrompt)
+_MessageParamT = TypeVar("_MessageParamT", bound=Any)
+_BaseDynamicConfigT = TypeVar("_BaseDynamicConfigT", bound=BaseDynamicConfig)
 
 
 class PromptDecorator(Protocol):
     @overload
-    def __call__(self, prompt: type[_BasePromptT]) -> type[_BasePromptT]: ...
+    def __call__(
+        self, prompt: Callable[_P, BaseDynamicConfig]
+    ) -> Callable[_P, list[BaseMessageParam]]: ...
 
     @overload
-    def __call__(self, prompt: Callable[_P, _R]) -> Callable[_P, _R]: ...
+    def __call__(
+        self, prompt: Callable[_P, Awaitable[BaseDynamicConfig]]
+    ) -> Callable[_P, Awaitable[list[BaseMessageParam]]]: ...
+
+    @overload
+    def __call__(self, prompt: type[_BasePromptT]) -> type[_BasePromptT]: ...
 
     def __call__(
         self,
-        prompt: type[_BasePromptT] | Callable[_P, _R],
-    ) -> type[_BasePromptT] | Callable[_P, _R]: ...
+        prompt: type[_BasePromptT]
+        | Callable[_P, BaseDynamicConfig]
+        | Callable[_P, Awaitable[BaseDynamicConfig]],
+    ) -> (
+        Callable[_P, list[BaseMessageParam]]
+        | Callable[_P, Awaitable[list[BaseMessageParam]]]
+        | type[_BasePromptT]
+    ): ...
+
+
+def _is_function(
+    prompt: type[_BasePromptT]
+    | Callable[_P, BaseDynamicConfig]
+    | Callable[_P, Awaitable[BaseDynamicConfig]],
+) -> TypeIs[
+    Callable[_P, BaseDynamicConfig] | Callable[_P, Awaitable[BaseDynamicConfig]]
+]:
+    return isinstance(prompt, types.FunctionType)
+
+
+def _get_prompt_messages_from_prompt_function(
```

Comment:
this function is not necessary since you can just use `get_fn_args` and `parse_prompt_messages` directly inside the decorator below